seed/php-sdk/imdb/src/Core/JsonDeserializer.php

```diff
@@ -0,0 +1,151 @@
+<?php
+
+namespace Seed\Core;
+
+use DateTime;
+use Exception;
+use JsonException;
+
+class JsonDeserializer
+{
+    /**
+     * @throws JsonException
+     */
+    public static function deserializeDateTime(string $datetime): DateTime
+    {
+        try {
+            return new DateTime($datetime);
+        } catch (Exception $e) {
+            throw new JsonException("Failed to create DateTime from string: $datetime", previous: $e);
+        }
+    }
+
+    /**
+     * @throws JsonException
+     */
+    public static function deserializeDate(string $date): DateTime
+    {
+        $dateTime = DateTime::createFromFormat(Constant::DateDeserializationFormat, $date);
+        if ($dateTime === false) {
+            throw new JsonException("Failed to create date from string: $date");
+        }
+        return $dateTime;
+    }
+
+    /**
+     * Deserializes the given array based on the type annotation.
+     *
+     * @param mixed[]|array<string, mixed> $data The array to be deserialized.
+     * @param mixed[]|array<string, mixed> $type The type definition from the annotation.
+     * @return mixed[]|array<string, mixed> The deserialized array.
+     */
+    public static function deserializeArray(array $data, array $type): array
+    {
+        return Utils::isMapType($type)
+            ? self::deserializeMap($data, $type)
+            : self::deserializeList($data, $type);
+    }
+
+    /**
+     * Deserializes a value based on the type definition.
+     *
+     * @param mixed $data The data to deserialize.
+     * @param mixed $type The type definition.
+     * @return mixed The deserialized value.
+     * @throws JsonException
+     */
+    private static function deserializeValue(mixed $data, mixed $type): mixed
```

Comment:
Does this need to support `object` / a class that implements `JsonSerializable`?

I noticed it's used in the `map` and `list` deserializers, so just thinking about the `map<string, User>` case.

Reply:
if the type is an `object`, it will fall through to `deserializeSingleValue` where it is handled